### ISSUE TITLE
Build failure on macOS 10.13 with CUDA 8.0 and 9.1

### ISSUE THREAD

**Issue by axman6** [AccelerateHS/accelerate-llvm#29]
I am submitting a...
- [x] bug report
- [ ] feature request
- [ ] support request => you might also like to ask your question on the [mailing list](mailto:accelerate-haskell@googlegroups.com) or [gitter chat](https://gitter.im/AccelerateHS/Lobby).

## Description
I've been trying to build accelerate-llvm and can't manage to get `cuda` to compile. `ld` can't find the `CUDA` framework when compiling `accelerate-llvm` using `stack --stack-yaml stack-8.2.yaml install`, but when checking out the `cuda` by itself and installing it with the same command works as expected. I originally has CUDA 8.0 installed and in trying find a solution also had the same problem with CUDA 9.1.

## Expected behaviour
A successful compile of accelerate-llvm which finds the CUDA framework.

## Current behaviour
Currently the following output is produces:

```
$ stack --stack-yaml stack-8.2.yaml install
accelerate-llvm-1.2.0.0: configure (lib)
accelerate-llvm-1.2.0.0: build (lib)
cuda-0.9.0.0: configure
cuda-0.9.0.0: build
accelerate-llvm-1.2.0.0: copy/register
Log files have been written to: /Users/mas17k/Haskell/accelerate-llvm/.stack-work/logs/
Progress: 2/5
--  While building custom Setup.hs for package cuda-0.9.0.0 using:
      /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/.stack-work/dist/x86_64-osx/Cabal-2.0.1.0/setup/setup --builddir=.stack-work/dist/x86_64-osx/Cabal-2.0.1.0 build --ghc-options " -ddump-hi -ddump-to-file -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/mas17k/Haskell/accelerate-llvm/.stack-work/logs/cuda-0.9.0.0.log

    [1 of 2] Compiling Main             ( /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/Setup.hs, /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/.stack-work/dist/x86_64-osx/Cabal-2.0.1.0/setup/Main.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/Setup.hs:344:43: warning: [-Wdeprecations]
        In the use of ‘die’ (imported from Distribution.Simple.Utils):
        Deprecated: "Messages thrown with die can't be controlled with Verbosity; use die' instead, or dieNoVerbosity if Verbosity truly is not available"
        |
    344 |             when (ldVersion < [2,25,1]) $ die $ windowsLinkerBugMsg ldPath
        |                                           ^^^
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/Setup.hs:404:16: warning: [-Wdeprecations]
        In the use of ‘die’ (imported from Distribution.Simple.Utils):
        Deprecated: "Messages thrown with die can't be controlled with Verbosity; use die' instead, or dieNoVerbosity if Verbosity truly is not available"
        |
    404 |     Nothing -> die cudaNotFoundMsg
        |                ^^^
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/Setup.hs:537:11: warning: [-Wdeprecations]
        In the use of ‘die’ (imported from Distribution.Simple.Utils):
        Deprecated: "Messages thrown with die can't be controlled with Verbosity; use die' instead, or dieNoVerbosity if Verbosity truly is not available"
        |
    537 |           die $ printf "Unexpected failure. Neither the default %s nor custom %s exist.\n" generatedBuildInfoFilePath customBuildInfoFilePath
        |           ^^^
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/Setup.hs:556:11: warning: [-Wdeprecations]
        In the use of ‘rawSystemProgramConf’
        (imported from Distribution.Simple.Program):
        Deprecated: "use runDbProgram instead"
        |
    556 |           rawSystemProgramConf verbosity c2hsProgram (withPrograms lbi) . filter (not . null) $
        |           ^^^^^^^^^^^^^^^^^^^^
    [2 of 2] Compiling StackSetupShim   ( /Users/mas17k/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/.stack-work/dist/x86_64-osx/Cabal-2.0.1.0/setup/StackSetupShim.o )
    Linking /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/.stack-work/dist/x86_64-osx/Cabal-2.0.1.0/setup/setup ...
    Configuring cuda-0.9.0.0...
    Found CUDA toolkit at: /Developer/NVIDIA/CUDA-9.1
    Storing parameters to cuda.buildinfo.generated
    Using build information from 'cuda.buildinfo.generated'.
    Provide a 'cuda.buildinfo' file to override this behaviour.
    Using build information from 'cuda.buildinfo.generated'.
    Provide a 'cuda.buildinfo' file to override this behaviour.
    Preprocessing library for cuda-0.9.0.0..
    Building library for cuda-0.9.0.0..
    [ 1 of 40] Compiling Foreign.CUDA.Internal.C2HS ( src/Foreign/CUDA/Internal/C2HS.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Internal/C2HS.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Internal/C2HS.hs:202:3: warning: [-Winline-rule-shadowing]
        Rule "cFloatConv/CFloat->Float" may never fire
          because ‘Foreign.C.Types.CFloat’ might inline first
        Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘Foreign.C.Types.CFloat’
        |
    202 |   "cFloatConv/CFloat->Float"   forall (x::Float).  cFloatConv CFloat x = x;
        |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Internal/C2HS.hs:204:3: warning: [-Winline-rule-shadowing]
        Rule "cFloatConv/CDouble->Double" may never fire
          because ‘Foreign.C.Types.CDouble’ might inline first
        Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘Foreign.C.Types.CDouble’
        |
    204 |   "cFloatConv/CDouble->Double" forall (x::Double). cFloatConv CDouble x = x
        |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    [ 2 of 40] Compiling Foreign.CUDA.Path ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Path.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Path.o )
    [ 3 of 40] Compiling Foreign.CUDA.Types ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Types.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Types.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Types.chs:111:16: warning: [-Wincomplete-patterns]
        Pattern match(es) are non-exhaustive
        In a case alternative: Patterns not matched: _ :: Int
        |
    111 | #ifdef USE_EMPTY_CASE
        |                ^^^^
    [ 4 of 40] Compiling Foreign.CUDA.Ptr ( src/Foreign/CUDA/Ptr.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Ptr.o )
    [ 5 of 40] Compiling Text.Show.Describe ( src/Text/Show/Describe.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Text/Show/Describe.o )
    [ 6 of 40] Compiling Foreign.CUDA.Runtime.Error ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Error.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Error.o )
    [ 7 of 40] Compiling Foreign.CUDA.Runtime.Utils ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Utils.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Utils.o )
    [ 8 of 40] Compiling Foreign.CUDA.Runtime.Texture ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Texture.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Texture.o )
    [ 9 of 40] Compiling Foreign.CUDA.Runtime.Stream ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Stream.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Stream.o )
    [10 of 40] Compiling Foreign.CUDA.Runtime.Marshal ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Marshal.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Marshal.o )
    [11 of 40] Compiling Foreign.CUDA.Runtime.Exec ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Exec.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Exec.o )
    [12 of 40] Compiling Foreign.CUDA.Runtime.Event ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Event.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Event.o )
    [13 of 40] Compiling Foreign.CUDA.Driver.Error ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Error.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Error.o )
    [14 of 40] Compiling Foreign.CUDA.Driver.Utils ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Utils.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Utils.o )
    [15 of 40] Compiling Foreign.CUDA.Driver.Stream ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Stream.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Stream.o )
    [16 of 40] Compiling Foreign.CUDA.Driver.Profiler ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Profiler.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Profiler.o )
    [17 of 40] Compiling Foreign.CUDA.Driver.Event ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Event.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Event.o )
    [18 of 40] Compiling Foreign.CUDA.Driver.IPC.Event ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/IPC/Event.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/IPC/Event.o )
    [19 of 40] Compiling Foreign.CUDA.Analysis.Device ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Analysis/Device.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Analysis/Device.o )
    [20 of 40] Compiling Foreign.CUDA.Runtime.Device ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Device.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime/Device.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Runtime/Device.chs:341:16: warning: [-Wincomplete-patterns]
        Pattern match(es) are non-exhaustive
        In a case alternative: Patterns not matched: _ :: Int
        |
    341 | instance Enum PeerFlag where
        |                ^^^^
    [21 of 40] Compiling Foreign.CUDA.Runtime ( src/Foreign/CUDA/Runtime.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Runtime.o )
    [22 of 40] Compiling Foreign.CUDA     ( src/Foreign/CUDA.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA.o )
    [23 of 40] Compiling Foreign.CUDA.Driver.Module.Base ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Base.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Base.o )
    [24 of 40] Compiling Foreign.CUDA.Driver.Module.Link ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Link.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Link.o )
    [25 of 40] Compiling Foreign.CUDA.Driver.Device ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Device.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Device.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Driver/Device.chs:71:16: warning: [-Wincomplete-patterns]
        Pattern match(es) are non-exhaustive
        In a case alternative: Patterns not matched: _ :: Int
       |
    71 |   {
       |                ^^^^
    [26 of 40] Compiling Foreign.CUDA.Driver.Context.Base ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Base.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Base.o )
    [27 of 40] Compiling Foreign.CUDA.Driver.Marshal ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Marshal.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Marshal.o )
    [28 of 40] Compiling Foreign.CUDA.Driver.Texture ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Texture.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Texture.o )
    [29 of 40] Compiling Foreign.CUDA.Driver.IPC.Marshal ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/IPC/Marshal.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/IPC/Marshal.o )
    [30 of 40] Compiling Foreign.CUDA.Driver.Context.Primary ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Primary.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Primary.o )
    [31 of 40] Compiling Foreign.CUDA.Driver.Context.Peer ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Peer.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Peer.o )
    
    /private/var/folders/fx/b6nz4cjj4kb0_zyq8r36b905vnctqw/T/stack6566/cuda-0.9.0.0/src/Foreign/CUDA/Driver/Context/Peer.chs:52:16: warning: [-Wincomplete-patterns]
        Pattern match(es) are non-exhaustive
        In a case alternative: Patterns not matched: _ :: Int
       |
    52 | #ifdef USE_EMPTY_CASE
       |                ^^^^
    [32 of 40] Compiling Foreign.CUDA.Driver.Context.Config ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Config.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context/Config.o )
    [33 of 40] Compiling Foreign.CUDA.Driver.Context ( src/Foreign/CUDA/Driver/Context.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Context.o )
    [34 of 40] Compiling Foreign.CUDA.Driver.Unified ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Unified.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Unified.o )
    [35 of 40] Compiling Foreign.CUDA.Driver.Exec ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Exec.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Exec.o )
    [36 of 40] Compiling Foreign.CUDA.Driver.Module.Query ( .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Query.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module/Query.o )
    [37 of 40] Compiling Foreign.CUDA.Driver.Module ( src/Foreign/CUDA/Driver/Module.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver/Module.o )
    [38 of 40] Compiling Foreign.CUDA.Driver ( src/Foreign/CUDA/Driver.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Driver.o )
    [39 of 40] Compiling Foreign.CUDA.Analysis.Occupancy ( src/Foreign/CUDA/Analysis/Occupancy.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Analysis/Occupancy.o )
    [40 of 40] Compiling Foreign.CUDA.Analysis ( src/Foreign/CUDA/Analysis.hs, .stack-work/dist/x86_64-osx/Cabal-2.0.1.0/build/Foreign/CUDA/Analysis.o )
    ld: framework not found CUDA
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    `gcc' failed in phase `Linker'. (Exit code: 1)
```

## Possible solution (optional)
None found yet

## Steps to reproduce (for bugs)

 1. Install latest CUDA (9.1)
 2. Clone this repo and cd into it
 3. run `stack --stack-yaml stack-8.2.yaml install`

## Your environment

- Version used: `git describe => 1.1.0.1-native-123-gd7db32a`
- Backend(s) used and version: All
- Operating system and version: macOS 10.13.2, CUDA 9.1
- Link to your project:
- If this is a bug with the GPU backend, include the output of `nvidia-device-query`: Probably not relevant, but included anyway (from the successfully built `cuda` repo):

```
$ stack exec nvidia-device-query
CUDA device query (Driver API, statically linked)
CUDA driver version 9.1
CUDA API version 9.1
Detected 1 CUDA capable device

Device 0: GeForce GT 750M
  CUDA capability:                    3.0
  CUDA cores:                         384 cores in 2 multiprocessors (192 cores/MP)
  Global memory:                      2 GB
  Constant memory:                    64 kB
  Shared memory per block:            48 kB
  Registers per block:                65536
  Warp size:                          32
  Maximum threads per multiprocessor: 2048
  Maximum threads per block:          1024
  Maximum grid dimensions:            2147483647 x 65535 x 65535
  Maximum block dimensions:           1024 x 1024 x 64
  GPU clock rate:                     925.5 MHz
  Memory clock rate:                  2.508 GHz
  Memory bus width:                   128-bit
  L2 cache size:                      256 kB
  Maximum texture dimensions         
    1D:                               65536
    2D:                               65536 x 65536
    3D:                               4096 x 4096 x 4096
  Texture alignment:                  512 B
  Maximum memory pitch:               2 GB
  Concurrent kernel execution:        Yes
  Concurrent copy and execution:      Yes, with 1 copy engine
  Runtime limit on kernel execution:  Yes
  Integrated GPU sharing host memory: No
  Host page-locked memory mapping:    Yes
  ECC memory support:                 No
  Unified addressing (UVA):           Yes
  PCI bus/location:                   1/0
  Compute mode:                       Default
    Multiple contexts are allowed on the device simultaneously
```

**Comment by tmcdonell** [comment]
Hi! I just upgraded to high sierra over the weekend, so good timing (;

I think I have fixed the `cuda` package now... try adding this to your `stack.yaml` and see if that works for you too?

```yaml
extra-deps:
- git:    https://github.com/tmcdonell/cuda.git
  commit: f04929808536fbcd28050d6bfd5ed682d8379d5b
```

**Comment by axman6** [comment]
Excellent timing! I hope you have less problems with High Sierra than I've had, the upgrade's been a nightmare for me. I'll give that a go and see how it goes.

**Comment by tmcdonell** [comment]
yeah, it took me ages to get everything up and running again too .___.

**Comment by axman6** [comment]
Success!

**Comment by tmcdonell** [comment]
Cool. Once I fix the build for 7.10 I'll upload a patch release.
